[threaded-animation-resolution] upload progress-based timelines through the remote layer tree

graouts opened the pull request:
#### 1e6f8e595987eb38a0b906594d96fa791dfb2d88
<pre>
[threaded-animation-resolution] upload progress-based timelines through the remote layer tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=301501">https://bugs.webkit.org/show_bug.cgi?id=301501</a>
<a href="https://rdar.apple.com/163469082">rdar://163469082</a>

Reviewed by Matt Woodrow.

In 301942@main we added a system to upload monotonic timelines from the web process to the remote
layer tree. In preparation for the support of scroll-driven animations running in the remote layer
tree, we add a new `ProgressResolutionData` struct to package all the relevant information for
progress-based timelines. The `m_originTime` member is now replaced with an `m_data` member which
holds a variant holding either monotonic (`Seconds`) or progress-based data (`ProgressResolutionData`).

The remote layer tree `RemoteAnimationTimeline` class now has two subclasses, `RemoteMonotonicTimeline`
and `RemoteProgressBasedTimeline`. The `RemoteAnimationTimelineRegistry` no longer exists and instead
two different registry classes are added, one for each timeline type: `RemoteMonotonicTimelineRegistry`
and `RemoteProgressBasedTimelineRegistry`. The former is backing monotonic timeline management for
`RemoteLayerTreeEventDispatcher` (macOS) and `RemoteScrollingCoordinatorProxyIOS` (iOS) while the latter
is backing progress-based timeline management for `RemoteScrollingTree` (both macOS and iOS).

Another change was required to adhere to our safe-programming guidelines and the `timeline()` lookup
methods now return a RefPtr instead of a raw pointer in order to avoid the use of `RefPtr::unsageGet()`
in `RemoteScrollingCoordinatorProxyIOS::timeline()` and `RemoteLayerTreeEventDispatcher::timeline()`.

Note that this patch does not enable the creation of remote progress-based timelines, nor does it add
the necessary code to update remote progress-based timelines and schedule invalidation of affected remote
layer tree nodes.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/WebAnimationTime.h:
* Source/WebCore/platform/animation/AcceleratedTimeline.cpp:
(WebCore::AcceleratedTimeline::create):
(WebCore::AcceleratedTimeline::AcceleratedTimeline):
(WebCore::AcceleratedTimeline::originTime const):
(WebCore::AcceleratedTimeline::progressResolutionData const):
* Source/WebCore/platform/animation/AcceleratedTimeline.h:
(WebCore::AcceleratedTimeline::isMonotonic const):
(WebCore::AcceleratedTimeline::isProgressBased const):
(WebCore::AcceleratedTimeline::data const):
(WebCore::AcceleratedTimeline::originTime const): Deleted.
* Source/WebCore/platform/animation/ProgressResolutionData.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.cpp:
(WebKit::RemoteAnimationTimeline::RemoteAnimationTimeline):
(WebKit::computeCurrentTime): Deleted.
(WebKit::RemoteAnimationTimeline::create): Deleted.
(WebKit::RemoteAnimationTimeline::updateCurrentTime): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.h:
(WebKit::RemoteAnimationTimeline::duration const):
(): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::timeline const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::timeline const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimeline.cpp: Copied from Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.cpp.
(WebKit::RemoteMonotonicTimeline::create):
(WebKit::RemoteMonotonicTimeline::RemoteMonotonicTimeline):
(WebKit::RemoteMonotonicTimeline::updateCurrentTime):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimeline.h: Copied from Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.h.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimelineRegistry.cpp: Renamed from Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimelineRegistry.cpp.
(WebKit::RemoteMonotonicTimelineRegistry::update):
(WebKit::RemoteMonotonicTimelineRegistry::get const):
(WebKit::RemoteMonotonicTimelineRegistry::advanceCurrentTime):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimelineRegistry.h: Copied from Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimelineRegistry.h.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimeline.cpp: Added.
(WebKit::RemoteProgressBasedTimeline::create):
(WebKit::RemoteProgressBasedTimeline::RemoteProgressBasedTimeline):
(WebKit::RemoteProgressBasedTimeline::updateCurrentTime):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimeline.h: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimelineRegistry.cpp: Added.
(WebKit::RemoteProgressBasedTimelineRegistry::update):
(WebKit::RemoteProgressBasedTimelineRegistry::get const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimelineRegistry.h: Added.
(WebKit::RemoteProgressBasedTimelineRegistry::isEmpty const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::timeline const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::updateTimelineRegistration):
(WebKit::RemoteScrollingTree::timeline const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateTimelineRegistration):
(WebKit::RemoteScrollingCoordinatorProxyIOS::timeline const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateAnimations):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.mm:
(WebKit::RemoteLayerTreeEventDispatcher::updateTimelineRegistration):
(WebKit::RemoteLayerTreeEventDispatcher::timeline):
(WebKit::RemoteLayerTreeEventDispatcher::updateAnimations):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::timeline const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/302282@main">https://commits.webkit.org/302282@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cfb2f66a6d893b9a4de112772cf91390e13b605b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128606 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/872 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39438 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135994 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80012 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/499a5030-4fb2-40a5-b53c-1a184064b6a8) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130478 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/816 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/746 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97900 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65807 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ed0bdcdb-9049-4c05-90f2-ac9da153faa1) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131554 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/603 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115226 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78517 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/01708b82-be6c-4004-871d-8ac555d00b4f) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33334 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79277 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108983 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33816 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138446 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/700 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/668 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106433 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/747 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111566 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106255 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/590 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30095 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53046 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20084 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/760 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63991 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/627 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/685 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/710 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.mm:
```diff
@@ -646,18 +650,24 @@ void displayLinkFired(PlatformDisplayID displayID, DisplayUpdate, bool wantsFull
     assertIsHeld(m_animationLock);
-    if (m_timelineRegistry)
-        return m_timelineRegistry->get(timelineID);
+    if (m_monotonicTimelineRegistry) {
+        if (RefPtr timeline = m_monotonicTimelineRegistry->get(timelineID))
+            return timeline.unsafeGet();
+    }
```
This is the newly-added code that is triggering the style bot error because we're introducing a new call to `RefPtr::unsafeGet()`. This seemed preferable to me over adding an exception for the use of an unprotected raw pointer, but I'm not sure that it is. Perhaps we should make this whole method and related one return `RefPtr<RemoteAnimationTimeline>`?

Cc @cdumez, @ddkilzer and @geoffreygaren who may have an opinion.

geoffreygaren replied:
Yes, I recommend returning RefPtr here. It is verifiably lifetime safe, and slightly more efficient.

No opinion on whether the callee function should change. Probably depends on whether it also contains a call to unsafeGet.

graouts replied:
Fixed in the latest iteration of this patch.

mattwoodrow commented on Source/WebCore/platform/animation/AcceleratedTimeline.h:
```diff
@@ -36,19 +37,29 @@ namespace WebCore {
     TimelineIdentifier m_identifier;
-    Seconds m_originTime;
+    std::optional<Seconds> m_originTime;
+    std::optional<ProgressResolutionData> m_progressResolutionData;
```
Variant, so that only one of these two can be valid?

graouts replied:
I considered it and got a bit lazy, but it's a better approach to ensure they're mutually exclusive. I'll update the patch to have a variant data member.

graouts replied:
This was addressed in the latest patch.

graouts committed:
[threaded-animation-resolution] upload progress-based timelines through the remote layer tree
https://bugs.webkit.org/show_bug.cgi?id=301501
rdar://163469082

Reviewed by Matt Woodrow.

In 301942@main we added a system to upload monotonic timelines from the web process to the remote
layer tree. In preparation for the support of scroll-driven animations running in the remote layer
tree, we add a new `ProgressResolutionData` struct to package all the relevant information for
progress-based timelines. The `m_originTime` member is now replaced with an `m_data` member which
holds a variant holding either monotonic (`Seconds`) or progress-based data (`ProgressResolutionData`).

The remote layer tree `RemoteAnimationTimeline` class now has two subclasses, `RemoteMonotonicTimeline`
and `RemoteProgressBasedTimeline`. The `RemoteAnimationTimelineRegistry` no longer exists and instead
two different registry classes are added, one for each timeline type: `RemoteMonotonicTimelineRegistry`
and `RemoteProgressBasedTimelineRegistry`. The former is backing monotonic timeline management for
`RemoteLayerTreeEventDispatcher` (macOS) and `RemoteScrollingCoordinatorProxyIOS` (iOS) while the latter
is backing progress-based timeline management for `RemoteScrollingTree` (both macOS and iOS).

Another change was required to adhere to our safe-programming guidelines and the `timeline()` lookup
methods now return a RefPtr instead of a raw pointer in order to avoid the use of `RefPtr::unsageGet()`
in `RemoteScrollingCoordinatorProxyIOS::timeline()` and `RemoteLayerTreeEventDispatcher::timeline()`.

Note that this patch does not enable the creation of remote progress-based timelines, nor does it add
the necessary code to update remote progress-based timelines and schedule invalidation of affected remote
layer tree nodes.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/WebAnimationTime.h:
* Source/WebCore/platform/animation/AcceleratedTimeline.cpp:
(WebCore::AcceleratedTimeline::create):
(WebCore::AcceleratedTimeline::AcceleratedTimeline):
(WebCore::AcceleratedTimeline::originTime const):
(WebCore::AcceleratedTimeline::progressResolutionData const):
* Source/WebCore/platform/animation/AcceleratedTimeline.h:
(WebCore::AcceleratedTimeline::isMonotonic const):
(WebCore::AcceleratedTimeline::isProgressBased const):
(WebCore::AcceleratedTimeline::data const):
(WebCore::AcceleratedTimeline::originTime const): Deleted.
* Source/WebCore/platform/animation/ProgressResolutionData.h: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.cpp:
(WebKit::RemoteAnimationTimeline::RemoteAnimationTimeline):
(WebKit::computeCurrentTime): Deleted.
(WebKit::RemoteAnimationTimeline::create): Deleted.
(WebKit::RemoteAnimationTimeline::updateCurrentTime): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.h:
(WebKit::RemoteAnimationTimeline::duration const):
(): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::timeline const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::timeline const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimeline.cpp: Copied from Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.cpp.
(WebKit::RemoteMonotonicTimeline::create):
(WebKit::RemoteMonotonicTimeline::RemoteMonotonicTimeline):
(WebKit::RemoteMonotonicTimeline::updateCurrentTime):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimeline.h: Copied from Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimeline.h.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimelineRegistry.cpp: Renamed from Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimelineRegistry.cpp.
(WebKit::RemoteMonotonicTimelineRegistry::update):
(WebKit::RemoteMonotonicTimelineRegistry::get const):
(WebKit::RemoteMonotonicTimelineRegistry::advanceCurrentTime):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimelineRegistry.h: Copied from Source/WebKit/UIProcess/RemoteLayerTree/RemoteAnimationTimelineRegistry.h.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimeline.cpp: Added.
(WebKit::RemoteProgressBasedTimeline::create):
(WebKit::RemoteProgressBasedTimeline::RemoteProgressBasedTimeline):
(WebKit::RemoteProgressBasedTimeline::updateCurrentTime):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimeline.h: Added.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimelineRegistry.cpp: Added.
(WebKit::RemoteProgressBasedTimelineRegistry::update):
(WebKit::RemoteProgressBasedTimelineRegistry::get const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteProgressBasedTimelineRegistry.h: Added.
(WebKit::RemoteProgressBasedTimelineRegistry::isEmpty const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
(WebKit::RemoteScrollingCoordinatorProxy::timeline const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::updateTimelineRegistration):
(WebKit::RemoteScrollingTree::timeline const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateTimelineRegistration):
(WebKit::RemoteScrollingCoordinatorProxyIOS::timeline const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::updateAnimations):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.mm:
(WebKit::RemoteLayerTreeEventDispatcher::updateTimelineRegistration):
(WebKit::RemoteLayerTreeEventDispatcher::timeline):
(WebKit::RemoteLayerTreeEventDispatcher::updateAnimations):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::timeline const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/302282@main
